app/src/main/java/com/example/sharewhatyoucanproject/donor/DonorViewModel.kt

```diff
@@ -0,0 +1,126 @@
+
+    private val _imageUrl = MutableLiveData<String>()
+    val imageUrl: LiveData<String> get() = _imageUrl
+
+    private val _saveData = MutableLiveData<String>()
+    val saveData: LiveData<String> get() = _saveData
+
+    var foodType: String = "Cooked Food"
+    var title = ""
+    var description = ""
+    var hourSet = ""
+    var filepath: Uri? = null
+
+    fun isDataCompleted(): Boolean {
+        return title.isNotEmpty() && description.isNotEmpty() && hourSet.isNotEmpty()
+    }
+
+    fun isImageNotSelected(): Boolean {
+        return filepath == null
+    }
+
+    fun isFoodNotValid(): Boolean {
+        val hours = Integer.parseInt(hourSet)
+        return foodType == "Cooked Food" && hours > 48 || foodType == "Groceries" && hours > 1460
+    }
+
+    fun saveData() {
+        val postMap: MutableMap<String, Any?> = HashMap()
+        postMap["imageUrl"] = imageUrl
+        postMap["date"] = FieldValue.serverTimestamp()
+        postMap["title"] = title
+        postMap["description"] = description
+        postMap["status"] = 0
+        postMap["type"] = foodType
+        postMap["location"] = currentLocation.value
+        db?.collection("posts")
+            ?.add(postMap)
+            ?.addOnCompleteListener { task ->
+                if (task.isSuccessful) {
+                    _saveData.value = "success"
+                } else {
+                    _saveData.value = "Failed ${task.exception}"
+                }
+            }
+    }
+
+    fun uploadImage() {
+        storageReference?.let {
+            val ref = it.child("images/" + UUID.randomUUID())
+            uploadTask = ref.putFile(filepath!!)
+            uploadTask.continueWithTask { task ->
+                if (!task.isSuccessful) {
+                    val excep = task.exception
+                    if (excep != null) {
+                        throw excep
+                    }
+                }
+                ref.downloadUrl
+            }.addOnCompleteListener { task ->
+                _imageUrl.value = task.result.toString()
+            }
+        }
+    }
+
+    fun getCurrentLocation() {
+        LocationServices.getFusedLocationProviderClient(context!!)
+            .requestLocationUpdates(
+                locationRequest,
+                object : LocationCallback() {
+                    override fun onLocationResult(locationresult: LocationResult) {
+                        super.onLocationResult(locationresult)
+                        if (locationresult.locations.size > 0) {
+                            val index: Int = locationresult.locations.size - 1
+                            val latitude = locationresult.locations[index].latitude
+                            val longitude = locationresult.locations[index].longitude
+                            _currentLocation.value = GeoPoint(latitude, longitude)
+                        } else {
+                            context.showToast("Some Technical Error. Please Retry")
```

Comment:
when you make your location lifecycle component, I recommend creating a callback to show this message instead of attempting to do it within the component